### ISSUE TITLE
Improve handling of punctuation in segmented languages

### ISSUE THREAD

**Issue by bglw** [CloudCannon/pagefind#352]
In whitespace-delimited languages, when Pagefind encounters `a-b` it will be indexed as `ab`. In languages that go through segmentation, this might have been first segmented to `'a', '-', 'b'` which indexes as the words `a` and `b` with a word in between. This makes it hard to search for, as on the client-side `a-b` will continue to search for `ab`, and an exact search for `"a b"` or `"a - b"` won't match due to the ignored word indexed between `a` and `b`.

We don't have the segmentation available on the client due to bandwidth constraints, so a different solution will need to be found. One easy(ish) option would be for the client to search for `a-b` as `ab` **and** `a b`.

**Comment by bglw** [comment]
Another option would be to swap out for different punctuation logic in segmented languages, so that in these cases `a-b` is always `'a', '-', 'b'`.

**Comment by hjonin** [comment]
Hi, I think the same problems occur with the symbol `'` that can be placed in front of a word in French (it's the contraction of the article `le`, which becomes `l'`).
For example, a search for `alphabet` should also return results in which `l'alphabet` can be found.
Thanks in advance and congratulations on a very good job!

**Comment by bglw** [comment]
Hey @hjonin 👋 

The French case is slightly different (in a good way) — this current issue only applies to non-whitespace-delimited languages like Chinese.

For indexing `l'alphabet`, that will be resolved by #225 which will be released on a stable version next week 😄

The current `v1.0.0-beta.2` release includes the behavior, though it currently _doesn't_ split on the `'` symbol. But you bring up a great point re:French, so I'll expand the logic to cover it (and probably just _all_ punctuation for now, and we can configure it down later if need be).

Here's an example of the new word indexing finding the `attribute` in `html_attribute`:

<img width="774" alt="Screenshot 2023-09-08 at 8 29 55 AM" src="https://github.com/CloudCannon/pagefind/assets/40188355/a6f4aa88-4f5c-440b-9a4c-0d7ed0cc2604">

**Comment by hjonin** [comment]
Hi @bglw 

Thank you very much for your answer! Can't wait to use the new version!

**Comment by remi-durand** [comment]
Hi @bglw
Hi @hjonin

**TL;DR;** Segmentation happens on sigle quote `'`  `U+0027` (HTML `&apos;`). Is there a way to "teach" pagefing that segmentation can also happens on `’` `U+2019` (HTML `&apos;`) using configuration?

**Details**

In French, it's common to use the the standard single quote as in `l'alphabet`. 
This quote is HTML `&apos;`, unicode `U+0027`.

But, for pretty printing, it's common to transform the single quote into another kind of quote: `l’alphabet`.
Here, the quote is HTML `&rsquo;`, unicode `U+2019`.

And as you can guess, some static website generators are performing this transformation behind the scenes. For instance, with HUGO, it's controlled with the configuration point : 

```
markup:
  goldmark:
    extensions:
      typographer:
        apostrophe: '&rsquo;'
```

And, since this a very common, this value is the default ([documentation](https://gohugo.io/configuration/markup/#goldmark)).

So, the issue is the following : with this standard transformation, segmentation does not perform as expected. Segmentation happens on `U+0027` but not with `U+2019`. Therefore, searching for `alphabet` matches  `l'alphabet` (which is expected), but not `l’alphabet` (which is unexpected).

Is there a way to "teach" pagefing that segmentation can also happens on `U+2019` using configuration?

**Comment by bglw** [comment]
Closing out as [v1.5.0](https://github.com/Pagefind/pagefind/releases/tag/v1.5.0) shipped a bunch of stuff in this area, so I strongly suspect many things in this area have been fixed. Do re-open if it's still seen on the latest Pagefind! (or preferably a new issue for maximum noise 😄)
